src/Common/SpaceSaving.h

```diff
@@ -308,88 +335,127 @@ class SpaceSaving
     {
         size_t alpha_size = 0;
         readVarUInt(alpha_size, rb);
+
+        size_t expected_capacity = alpha_map.size();
+        if (alpha_size != expected_capacity)
+            throw DB::Exception(
+                DB::ErrorCodes::SIZES_OF_ARRAYS_DONT_MATCH,
+                "Found incorrect alpha vector size (Passed: {}. Expected: {})",
+                alpha_size,
+                expected_capacity);
+
+        for (size_t i = 0; i < alpha_size; ++i)
+        {
+            UInt64 alpha = 0;
+            readVarUInt(alpha, rb);
+            alpha_map[i] = alpha;
+        }
+    }
+
+    void skipAlphaMap(ReadBuffer & rb)
+    {
+        size_t alpha_size = 0;
+        readVarUInt(alpha_size, rb);
+
         for (size_t i = 0; i < alpha_size; ++i)
         {
             UInt64 alpha = 0;
             readVarUInt(alpha, rb);
-            alpha_map.push_back(alpha);
         }
     }
 
 protected:
-    void push(std::unique_ptr<Counter> counter)
+    NO_INLINE void push(Counter && counter)
     {
-        counter->slot = counter_list.size();
-        auto * ptr = counter.get();
+        size_t pos = counter_list.size();
+        counter_map.insertIfNotPresent(counter.key, counter.hash, pos);
         counter_list.push_back(std::move(counter));
-        counter_map[ptr->key] = ptr;
-        percolate(ptr);
+        truncateIfNeeded(false);
     }
 
-    // This is equivalent to one step of bubble sort
-    void percolate(Counter * counter)
+    ALWAYS_INLINE void truncateIfNeeded(bool force_counter_map_rebuild)
     {
-        while (counter->slot > 0)
+        if (unlikely(counter_list.size() >= capacity() * 2))
         {
-            auto & next = counter_list[counter->slot - 1];
-            if (*counter > *next)
+            ::nth_element(
+                counter_list.begin(),
+                counter_list.begin() + capacity(),
+                counter_list.end(),
+                [](const auto & l, const auto & r) { return l > r; });
+
+            const UInt64 alpha_mask = alpha_map.size() - 1;
+            for (size_t i = m_capacity; i < counter_list.size(); ++i)
             {
-                std::swap(next->slot, counter->slot);
-                std::swap(counter_list[next->slot], counter_list[counter->slot]);
+                size_t pos = counter_list[i].hash & alpha_mask;
+                alpha_map[pos] = std::min(alpha_map[pos] + counter_list[i].count - counter_list[i].error, MAX_ALPHA_VALUE);
+                arena.free(counter_list[i].key);
             }
-            else
-                break;
+
+            counter_list.resize(m_capacity);
+            force_counter_map_rebuild = true;
+        }
+
+        if (force_counter_map_rebuild)
+        {
+            /// Rebuild the counter map
+            counter_map.clear();
```

Comment:
Do you have an idea how `ClearableHashMap` would perform compared to the normal one?

Reply:
Tested and it's faster. I'll push it